### PR TITLE
Release notes for Percona Server 5.6.32-78.0 (5.6)

### PR DESCRIPTION
**Pull request by hrvojem**

### CODE REVIEW

File: doc/source/management/audit_log_plugin.rst [hunk 1]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+In :rn:`5.6.32-78.0` |Percona Server| has implemented filtering by user. This
+was implemented by adding two new global variables:
+:variable:`audit_log_include_accounts` and
+:variable:`audit_log_exclude_accounts` to specify which user accounts should be
+included or excluded from audit logging. 
+
+.. warning:: 
+
+  Only one of these variables can contain the list of users to be either
+  included or exclude, while other needs to be ``NULL``. If one of the
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: s/exclude/excluded, s/other/the other

File: doc/source/management/audit_log_plugin.rst [hunk 2]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+  variables is set to be not ``NULL`` (contains a list of users), the attempt
+  to set another one will fail. Empty string means an empty list.
+
+.. note::
+
+  Change of :variable:`audit_log_include_accounts` and
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: Changes of ... do not apply to existing server connections.

File: doc/source/management/audit_log_plugin.rst [hunk 3]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+
+If you you try to add users to both include and exclude lists server will show
+you the following error:
+
+.. code-block:: mysql
+
+  mysql> SET GLOBAL audit_log_exclude_accounts = 'user1@localhost,root@localhost';
+  ERROR 1231 (42000): Variable 'audit_log_exclude_accounts' can't be set to the value of 'user1@localhost,root@localhost'
+
+You can add users to a different list by first setting the value of list they
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: To switch from filtering by included user list to the excluded one or back, first set the currently active filtering variable to NULL:

File: doc/source/management/audit_log_plugin.rst [hunk 4]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+.. code-block:: mysql
+
+  SET GLOBAL audit_log_exclude_accounts = 'user1@localhost';
+
+The value can be ``NULL`` or comma separated list of accounts in form
+``user@host`` or ``'user'@'host'`` (if user or host contains comma).
+
+.. _filtering_by_sql_command:
+
+Filtering by SQL command
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: command type

File: doc/source/management/audit_log_plugin.rst [hunk 5]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+========================
+
+In :rn:`5.6.32-78.0` |Percona Server| has implemented filtering by SQL command.
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: "s/SQL command/SQL command type" everywhere where appropriate

File: doc/source/management/audit_log_plugin.rst [hunk 6]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+This was implemented by adding two new global variables:
+:variable:`audit_log_include_commands` and
+:variable:`audit_log_exclude_commands` to specify which commands should be
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: command types

File: doc/source/management/audit_log_plugin.rst [hunk 7]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+included or excluded from audit logging.
+
+.. warning:: 
+
+  Only one of these variables can contain the list of users to be either
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: s/users/command types/g

File: doc/source/management/audit_log_plugin.rst [hunk 8]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+  included or exclude, while other needs to be ``NULL``. If one of the
+  variables is set to be not ``NULL`` (contains a list of users), the attempt
+  to set another one will fail. Empty string means an empty list.
+
+.. note:: 
+
+  If both :variable:`audit_log_exclude_accounts` and
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: Variable name

File: doc/source/management/audit_log_plugin.rst [hunk 9]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+  :variable:`audit_log_include_accounts` are ``NULL`` every session will be
+  logged.
+
+Example
+-------
+
+To check the available commands you can run:
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: The available command types can be listed by running

File: doc/source/management/audit_log_plugin.rst [hunk 10]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+    STATUS="0"
+    SQLTEXT="create database world"
+    USER="root[root] @ localhost []"
+    HOST="localhost"
+    OS_USER=""
+    IP=""
+    DB=""
+  />
+
+You can add commands to a different list by first setting the value of list they are currently in to ``NULL``:
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: To switch command type filtering type from included type list to excluded one or back, first reset the currently-active list to NULL:

File: doc/source/management/audit_log_plugin.rst [hunk 11]
```diff
@@ -141,18 +141,229 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+
+  mysql> SET GLOBAL audit_log_include_commands = NULL;
+  Query OK, 0 rows affected (0.00 sec)
+
+  mysql> SET GLOBAL audit_log_exclude_commands= 'set_option,create_db';
+  Query OK, 0 rows affected (0.00 sec)
+
+.. note::
+
+  Invocation of stored procedure has ``sql_command`` field ``call_procedure``. All statements within the procedure have the ``sql_command`` ``call_procedure`` as well.
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: Invocation of stored procedures have command type call_procedure, and all the statements executed within the procedure have the same type call_procedure as well.

File: doc/source/management/audit_log_plugin.rst [hunk 12]
```diff
@@ -196,28 +407,68 @@ When this variable is set to ``ON`` log file will be closed and reopened. This c
 
 This variable can be used to specify the size of memory buffer used for logging, used when :variable:`audit_log_strategy` variable is set to ``ASYNCHRONOUS`` or ``PERFORMANCE`` values. This variable has effect only when :variable:`audit_log_handler` is set to ``FILE``.
 
+.. variable:: audit_log_exclude_accounts
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst:414]: include?

File: doc/source/management/audit_log_plugin.rst [hunk 13]
```diff
@@ -196,28 +407,68 @@ When this variable is set to ``ON`` log file will be closed and reopened. This c
...
+
+.. variable:: audit_log_exclude_commands
+
+    :version 5.6.38-78.0: Implemented
+    :cli: Yes
+    :scope: Global
+    :dyn: Yes
+    :vartype: String
+
+This variable is used to specify the list of commands for which :ref:`filtering_by_user` is applied. The value can be ``NULL`` or comma separated list of commands.
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: form should be described here too?

File: doc/source/management/audit_log_plugin.rst [hunk 14]
```diff
@@ -196,28 +407,68 @@ When this variable is set to ``ON`` log file will be closed and reopened. This c
...
 
+.. variable:: audit_log_exclude_accounts
+
+    :version 5.6.38-78.0: Implemented
+    :cli: Yes
+    :scope: Global
+    :dyn: Yes
+    :vartype: String
+
+This variable is used to specify the list of users for which :ref:`filtering_by_user` is applied. The value can be ``NULL`` or comma separated list of accounts in form ``user@host`` or ``'user'@'host'`` (if user or host contains comma).
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: If this variable is set, then audit_log_exclude_commands must be unset, and vice versa. - here and below

File: doc/source/management/audit_log_plugin.rst [hunk 15]
```diff
@@ -196,28 +407,68 @@ When this variable is set to ``ON`` log file will be closed and reopened. This c
...
+    :cli: Yes
+    :scope: Global
+    :dyn: No 
+    :vartype: String
+    :default: OLD
+    :allowed values: ``OLD``, ``NEW``, ``CSV``, ``JSON``
 
 This variable is used to specify the audit log format. The audit log plugin supports four log formats: ``OLD``, ``NEW``, ``JSON``, and ``CSV``. ``OLD`` and ``NEW`` formats are based on XML, where the former outputs log record properties as XML attributes and the latter as XML tags. Information logged is the same in all four formats.
 
+.. variable:: audit_log_include_accounts
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst:454]: users

File: doc/source/management/audit_log_plugin.rst [hunk 16]
```diff
@@ -196,28 +407,68 @@ When this variable is set to ``ON`` log file will be closed and reopened. This c
...
+
+    :version 5.6.38-78.0: Implemented
+    :cli: Yes
+    :scope: Global
+    :dyn: Yes
+    :vartype: String
+
+This variable is used to specify the list of users for which :ref:`filtering_by_user` is applied. The value can be ``NULL`` or comma separated list of accounts in form ``user@host`` or ``'user'@'host'`` (if user or host contains comma).
+
+.. variable:: audit_log_include_commands
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst:468]: exclude users

File: doc/source/management/audit_log_plugin.rst [hunk 17]
```diff
@@ -196,28 +407,68 @@ When this variable is set to ``ON`` log file will be closed and reopened. This c
...
+
+    :version 5.6.38-78.0: Implemented
+    :cli: Yes
+    :scope: Global
+    :dyn: Yes
+    :vartype: String
+
+This variable is used to specify the list of commands for which :ref:`filtering_by_user` is applied. The value can be ``NULL`` or comma separated list of commands.
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: All account filtering variable comments apply too

File: doc/source/release-notes/Percona-Server-5.6.32-78.0.rst [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+ |Percona Server| :ref:`audit_log_plugin` now supports filtering by :ref:`user
+ <filtering_by_user>` and :ref:`sql_command <filtering_by_sql_command>`.
+
+ |Percona Server| now supports tree map file block allocation strategy for
+ |TokuDB|.
+
+Bugs Fixed
+==========
+
+ Fixed potential cardinality ``0`` issue if ``ANALYZE TABLE`` finds only
```
Comment (laurynas-biveinis) [inline doc/source/release-notes/Percona-Server-5.6.32-78.0.rst]: Only for TokuDB tables?

File: doc/source/management/audit_log_plugin.rst (hunk 12 above)
Comment (hrvojem) [inline doc/source/management/audit_log_plugin.rst:414]: variables are sorted alphabetically

File: doc/source/management/audit_log_plugin.rst (hunk 15 above)
Comment (hrvojem) [inline doc/source/management/audit_log_plugin.rst:454]: variables are sorted alphabetically

File: doc/source/management/audit_log_plugin.rst (hunk 16 above)
Comment (hrvojem) [inline doc/source/management/audit_log_plugin.rst:468]: variables are sorted alphabetically

File: doc/source/management/audit_log_plugin.rst (hunk 12 above)
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst:414]: No, I mean it's audit_log_exclude_accounts twice? copy-paste error?

File: doc/source/management/audit_log_plugin.rst (hunk 16 above)
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst:468]: Likewise two audit_log_include_commands variables

File: doc/source/management/audit_log_plugin.rst [hunk 19]
```diff
@@ -141,18 +141,233 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+type. This was implemented by adding two new global variables:
+:variable:`audit_log_include_commands` and
+:variable:`audit_log_exclude_commands` to specify which command types should be
+included or excluded from audit logging.
+
+.. warning:: 
+
+  Only one of these variables can contain the list of command types to be
+  either included or excluded, while the other needs to be ``NULL``. If one of
+  the variables is set to be not ``NULL`` (contains a list of users), the attempt
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: a list of command types

File: doc/source/management/audit_log_plugin.rst [hunk 20]
```diff
@@ -141,18 +141,233 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+  to set another one will fail. Empty string means an empty list.
+
+.. note:: 
+
+  If both :variable:`audit_log_exclude_commands` and
+  :variable:`audit_log_include_commands` are ``NULL`` every session will be
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: all command will be logged

File: doc/source/release-notes/Percona-Server-5.6.32-78.0.rst [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+ |TokuDB| ``database.table.index`` names longer than 256 characters could cause
+ server crash if :ref:`background analyze table
+ <tokudb_background_analyze_table>` status was checked while running. Bug fixed
+ :tokubug:`1005`.
+
+ :ref:`pam_plugin` would abort authentication while checking UNIX user group
+ membership if there were more than a thousand members. Bug fixed
+ :bug:`1608902`.
+
+ Fixed reading past the end of heap buffer on saving the default DB in
```
Comment (laurynas-biveinis) [inline doc/source/release-notes/Percona-Server-5.6.32-78.0.rst]: other bugs

File: doc/source/management/audit_log_plugin.rst [hunk 22]
```diff
@@ -141,18 +141,233 @@ Ability to stream the audit log to `syslog <http://www.syslog.org/>`_ was implem
...
+type. This was implemented by adding two new global variables:
+:variable:`audit_log_include_commands` and
+:variable:`audit_log_exclude_commands` to specify which command types should be
+included or excluded from audit logging.
+
+.. warning:: 
+
+  Only one of these variables can contain a list of command types to be
+  either included or excluded, while the other needs to be ``NULL``. If one of
+  the variables is set to be not ``NULL`` (contains a list of users), the attempt
```
Comment (laurynas-biveinis) [inline doc/source/management/audit_log_plugin.rst]: a list of command types

**Comment by george-lorch** [comment]
TokuDB bits +1

**Pull request merged**